### PR TITLE
Add `yarn-plugin-backstage`

### PR DESCRIPTION
**Pull request by mtlewis**
## Hey, I just made a Pull Request!

Adds the yarn plugin proposed in https://github.com/backstage/backstage/issues/24065. Some outstanding discussions to resolve (not necessarily before shipping, but before we advocate for widespread adoption):

### Outstanding todos

* Tests for resolver - be nice to have an integration test here in particular that verifies the behaviour of the resolver by running yarn directly.

### What version range do we recommend?

Right now, this plugin resolves the backstage version to a single concrete version for each package, **with no prefix on the range**. My rationale is this: if backstage.json always contains a single Backstage version, and if our plugin is reliably mapping that version to the right version of each package, then there's no reason to be permissive about the version range. This is of course a deviation from our current standard, where packages are generally installed by create-app with a `^` prefix.

### How do users access the plugin?

Yarn plugins are not accessed via npm. Instead, they must be pulled from a remote URL. It seems like there's two common patterns for where plugins can live:
1. Colocated with the code on github (like [this one](https://github.com/mhassan1/yarn-plugin-after-install?tab=readme-ov-file#install)). This is the easiest, but also likely a bit messier to maintain, since whenever we release we need to commit back to the repo.
2. Hosted on a website - like [this](https://yarn.build/). If there's a convenient way for us to make this plugin available somewhere on backstage.io, this might be a nicer option from a UX perspective. 

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by Rugvip — COMMENTED** [review]
Oh interesting that that is how installation works, what about updates? I'm using that's be a fairly manual step too? Could you perhaps add a short README?

One thought is whether this should live in a different repo for simplicity

**Comment by mtlewis** [comment]
> Oh interesting that that is how installation works, what about updates? I'm using that's be a fairly manual step too? 

As with installation, users have to do it manually. Have been thinking over this a little bit, and wondering whether we might want to bake updating the plugin (or even prompting the user to install it and migrate to `backstage:*` versions) into `versions:bump`. I think this is a few steps down the road though, in my view the first step here should be getting the plugin published somewhere so we can have a few folks try it out and see if there are gaps.

> Could you perhaps add a short README?

There's a readme in the PR, but it doesn't have installation docs - instead, it just asserts that this isn't ready for use yet, since it's not properly accounted for in the ecosystem (in particular, I'd expect running `versions:bump` to clobber these versions right now). WDYT about keeping the readme light until we're ready to recommend folks adopt?

> One thought is whether this should live in a different repo for simplicity

Totally. Had the same thought. I think it comes down to what release cadence we want for this. It seems to me it's either:

1. We release it ad-hoc when we need it to change - and this means that we bake in a single version of the @backstage/release-manifests and don't update it with core releases. If we do this, I think a separate repo makes sense.
2. We release it along with overall Backstage releases, and include the ability to pull the plugin bundle from a specific release, or the release lines. In this case it probably makes sense to keep it in the Backstage repo, and weave the appropriate publishing work into the release workflow.

Wondering if it makes sense to remove the bundle code from this PR now, since it pushes us down a specific path for publishing.

**Comment by mtlewis** [comment]
> Wondering if it makes sense to remove the bundle code from this PR now, since it pushes us down a specific path for publishing.

This is now done - thoughts on merging this then following up with work to get the plugin published somewhere consumable?

**Comment by mtlewis** [comment]
Added the pre-pack step to insert the package version - and have also restricted this plugin to _only_ support `"backstage:^"`, because I think this is what we want everyone to use. Now this is done, this is ready for review.

**Comment by mtlewis** [comment]
@Rugvip this is ready for final review from my perspective :+1:.

File: packages/yarn-plugin/src/util.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  return false;
+};
+
+const findWorkspaceRoot = () => {
+  const cwd = ppath.cwd();
+  let currentDir = cwd;
+
+  while (!isWorkspaceRoot(currentDir)) {
```
Comment (Rugvip) [inline packages/yarn-plugin/src/util.ts:41]: Think it would be good to add a limit to this, just in case

File: packages/yarn-plugin/src/util.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  const manifest = await getManifestByVersion({
+    version: range.selector,
+  });
+
+  const manifestEntry = manifest.packages.find(
+    candidate => candidate.name === ident,
+  );
+
+  if (!manifestEntry) {
+    throw new Error(`Package ${ident} not found in manifest`);
```
Comment (Rugvip) [inline packages/yarn-plugin/src/util.ts:89]: Thinking we might want to improve the error messaging here with a bit more guidance and explanation. In particular, mentioning that the package is not part of a particular Backstage release. We could probably guide users toward the potential need to replace the package too.

**Review by Rugvip — CHANGES_REQUESTED** [review]
Nice! 👍 

Just a couple of nits

File: packages/yarn-plugin/src/util.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    if (manifest.workspaces) {
+      return true;
+    }
+  }
+
+  return false;
+};
+
+const findWorkspaceRoot = () => {
```
Comment (Rugvip) [inline packages/yarn-plugin/src/util.ts:37]: Explore using `cli-common` here instead. It's a very lightweight package

File: packages/yarn-plugin/src/resolver/BackstageResolver.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      );
+    }
+
+    return structUtils.makeDescriptor(
+      descriptor,
+      `${PROTOCOL}${getCurrentBackstageVersion()}`,
+    );
+  }
+
+  async getCandidates(descriptor: Descriptor): Promise<Locator[]> {
```
[the remark this answers is not in the document]
Comment (mtlewis) [reply packages/yarn-plugin/src/resolver/BackstageResolver.ts:69]: Done in acc7caa7151b88c21a6a1f986e034b8ad534c420

File: packages/yarn-plugin/src/util.ts (hunk 1 above)
Comment (mtlewis) [reply packages/yarn-plugin/src/util.ts]: Superseded by https://github.com/backstage/backstage/pull/24136#discussion_r1625698129

File: packages/yarn-plugin/src/util.ts (hunk 2 above)
Comment (mtlewis) [reply packages/yarn-plugin/src/util.ts]: Done in 95d7234e91392e1ff8feb0ba7e473658e2f3cb2a

File: packages/yarn-plugin/src/util.ts (hunk 3 above)
Comment (mtlewis) [reply packages/yarn-plugin/src/util.ts]: Done in 573e98b352484183a8c96e37aad7f771bcd9235c

**Pull request merged**